markdown/ch6.markdown

```diff
@@ -697,63 +798,62 @@ $$
 $$
 
-We can solve this equation using `scipy.sparse`'s iterative
-*biconjugate gradient (bicg)* solver^[bicgstab].  (Note that, if you
-had a symmetric matrix, you would use Conjugate Gradients,
-``scipy.sparse.linalg.solve.cg``, instead.)
-
-[^bicgstab]: The default formulation for bicg iteration is numerically
-    unstable, so we use the stabilized variation, which replaces every
-    second iteration with a *generalized minimal residual method*
-    (GMRES) iteration.
+We can solve this equation using `scipy.sparse.linalg`'s direct
+solver, `spsolve`. Depending on the structure and size of a linear algebra
+problem, though, it might be more efficient to use an iterative solver. See
+the [`scipy.sparse.linalg` documentation](http://docs.scipy.org/doc/scipy/reference/sparse.linalg.html#solving-linear-problems)
+for more information on this.
 
 ```python
-from scipy.sparse.linalg.isolve import bicgstab as bicg
+from scipy.sparse.linalg import spsolve
 
-damping = 0.99
+damping = 0.85
+beta = 1 - damping
 
 I = sparse.eye(n, format='csc')  # Same sparse format as Trans
 
-pagerank, error = bicg(I - damping * Trans,
-                       (1-damping) / n * np.ones(n),
-                       maxiter=int(1e4))
-print('error code: ', error)
+pagerank = spsolve(I - damping * Trans,
+                   np.full(n, beta / n))
 ```
 
-As can be seen in the documentation for the `bicg` solver, an error code of 0
-indicates that a solution was found! We now have the "foodrank" of the
-St. Marks food web!
+We now have the "foodrank" of the St. Marks food web!
 
 So how does a species' foodrank compare to the number of other species eating
 it?
 
 ```python
-def pagerank_plot(names, in_degrees, pageranks,
-                  annotations=[]):
-    fig, ax = plt.subplots(figsize=(9, 5))
+def pagerank_plot(in_degrees, pageranks, names, *,
+                  annotations=[], **figkwargs):
+    """Plot node pagerank against in-degree, with hand-picked node names."""
+
+    fig, ax = plt.subplots(**figkwargs)
     ax.scatter(in_degrees, pageranks, c=[0.835, 0.369, 0], lw=0)
-    labels = []
     for name, indeg, pr in zip(names, in_degrees, pageranks):
         if name in annotations:
             text = ax.text(indeg + 0.1, pr, name)
-            labels.append(text)
+
     ax.set_ylim(0, np.max(pageranks) * 1.1)
     ax.set_xlim(-1, np.max(in_degrees) * 1.1)
     ax.set_ylabel('PageRank')
     ax.set_xlabel('In-degree')
+```
+
+We now draw the plot. Having explored the dataset before writing this, we have
+pre-labeled some interesting nodes in the plot:
 
+```python
 interesting = ['detritus', 'phytoplankton', 'benthic algae', 'micro-epiphytes',
-               'microfauna', 'zooplankton', 'predatory shrimps', 'meiofauna', 'gulls']
+               'microfauna', 'zooplankton', 'predatory shrimps', 'meiofauna',
+               'gulls']
 in_degrees = np.ravel(Adj.sum(axis=0))
 pagerank_plot(species, in_degrees, pagerank, annotations=interesting)
```

Comment:
Should be:
`pagerank_plot(in_degrees, pagerank, species, annotations=interesting)`